parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -212,8 +212,20 @@ private static MetadataFilter filter(boolean skipRowGroups) {
   }
 
   private static <T> List<T> runAllInParallel(int parallelism, List<Callable<T>> toRun) throws ExecutionException {
-    LOG.info("Initiating action with parallelism: {}", parallelism);
-    ExecutorService threadPool = Executors.newFixedThreadPool(parallelism);
+    if (toRun.isEmpty()) {
+      return Collections.emptyList();
+    } else if (toRun.size() == 1) {
+      // Prevent instantiating a ThreadPool if only one task
+      try {
+		return Collections.singletonList(toRun.get(0).call());
+      } catch (Exception e) {
+        throw new ExecutionException(e);
+      }
+    }
+    // Prevent instantiating a pool with many threads if not useful
+    int limitedParallelism = Math.min(parallelism, toRun.size());
```

Comment:
I think this is a reasonable and safe optimization. There is no need to create a threadpool larger than the number of callables to run. However, I also think that it does not necessarily make sense to create that many either. For example, if there are 1000 callables, I would not create a separate thread for all of them. I think this number should be limited on the other boundary as well, i.e., it should not be more than `Runtime.getRuntime().availableProcessors()`.